File: haskell/src/Solutions/Euler187.hs
module Solutions.Euler187 where
import Utils.NumberTheory(unPrime, primes)
import Utils.Operators((<&&>))

ps = map unPrime primes :: [Int]

bound = 10^8
f a = length $ takeWhile ((<=a) <&&> (<bound) . (*a)) ps

euler187 :: IO String
euler187 = return $ show $ sum $ map f $ takeWhile (<bound `div` 2) ps

File: rust/src/solutions/euler187.rs
pub fn euler187() -> String {
    let bound = 100_000_000;
    let sieve = primal::Sieve::new(bound);
    sieve
        .primes_from(0)
        .take_while(|&p| p < bound / 2)
        .map(|p| {
            sieve
                .primes_from(0)
                .take_while(|&pp| pp <= p && pp * p < bound)
                .count()
        })
        .sum::<usize>()
        .to_string()
}
